### PR TITLE
Fix bug/code smell where GA cookie is set after a server redirect

### LINKED ISSUE #250: Fix bug where GA cookie is set after a server redirect
**Describe the bug**
When a new visitor without a GA cookie visits a short link, the server attempts to set a cookie after a response has already taken place. This results in a HTTP 500 Internal Server Error being logged when in fact a HTTP 302 Redirect has already been returned to the client.

**To reproduce**
1. Clear cookies
2. Visit a short link
3. Verify that browser is redirected successfully
4. Observe stack trace and HTTP 500 in logs

### PR DESCRIPTION
**Pull request by yong-jie**
## Problem

When a new visitor without a GA cookie visits a short link, the server attempts to set a cookie after a response has already taken place. This results in a HTTP 500 Internal Server Error being logged when in fact a HTTP 302 Redirect has already been returned to the client.

Closes #250

## Solution

- [x] Move cookie generation and setting to another place before setting res
- [ ] Don't create a new `logRedirect` function in every redirect
- [ ] Don't pass res and req into other functions
- [x] Fix broken tests

### CODE REVIEW

File: src/server/controllers/RedirectController.ts [hunk 1]
```diff
@@ -56,12 +56,18 @@ export class RedirectController implements RedirectControllerInterface {
...
           longUrl,
         )
       }
 
+      const generatedCookie = this.analyticsLogger.generateCookie(
+        req.headers.cookie,
+      )
+      if (generatedCookie) {
+        res.cookie(...generatedCookie)
+      }
```
Comment (LoneRifle) [inline src/server/controllers/RedirectController.ts:69]: The logic here is a bit roundabout - basically we are saying "speculatively generate the cookie, and if we happen to have one, set it". It really should say "check for the presence of a cookie, and if absent, generate one and set it"

File: src/server/services/analyticsLogger.ts [hunk 2]
```diff
@@ -13,25 +13,30 @@ export interface AnalyticsLogger {
...
+
+  generateCookie: (
+    cookie?: string,
+  ) => [string, string, { maxAge: number }] | null
 }
 
 @injectable()
 export class GaLogger implements AnalyticsLogger {
   logRedirectAnalytics: (
     req: Express.Request,
```
Comment (liangyuanruo) [inline src/server/services/analyticsLogger.ts:28]: can we refactor out the code smell with the need for `req` here?

Comment (LoneRifle) [reply src/server/services/analyticsLogger.ts:28]: Given the turnaround time, let's hold off on that. I can refactor this after the release

**Comment by yong-jie** [comment]
Acknowledged that code refactors are not the priority in this PR, and will be addressed in a separate PR after today's release.

**Pull request merged**